File: src/modules/wishlist/wishlist.controller.ts
import type { Request, Response } from 'express';
import { AppError } from '../../shared/errors/AppError.js';
import { sendSuccess } from '../../shared/utils/response.js';
import * as wishlistService from './wishlist.service.js';

function requireUser(req: Request): { userId: string } {
  if (!req.user) throw AppError.unauthorized('Authentication required');
  return { userId: req.user.userId };
}

export async function getWishlist(req: Request, res: Response): Promise<void> {
  const { userId } = requireUser(req);
  const items = await wishlistService.getWishlist(userId);
  sendSuccess(res, items, 'Wishlist retrieved successfully');
}

export async function addToWishlist(req: Request, res: Response): Promise<void> {
  const { userId } = requireUser(req);
  const { productId } = req.params as { productId: string };
  const items = await wishlistService.addToWishlist(userId, productId);
  sendSuccess(res, items, 'Item added to wishlist successfully');
}

export async function removeFromWishlist(req: Request, res: Response): Promise<void> {
  const { userId } = requireUser(req);
  const { productId } = req.params as { productId: string };
  const items = await wishlistService.removeFromWishlist(userId, productId);
  sendSuccess(res, items, 'Item removed from wishlist');
}

export async function clearWishlist(req: Request, res: Response): Promise<void> {
  const { userId } = requireUser(req);
  const items = await wishlistService.clearWishlist(userId);
  sendSuccess(res, items, 'Wishlist cleared');
}

File: java/src/main/java/com/ecommerce/backend/modules/wishlist/WishlistController.java
package com.ecommerce.backend.modules.wishlist;

import com.ecommerce.backend.common.api.ApiResponse;
import com.ecommerce.backend.common.error.ApiException;
import com.ecommerce.backend.modules.wishlist.WishlistService.WishlistItemPublic;
import com.ecommerce.backend.security.CurrentUser;
import java.util.List;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

/** Wishlist endpoints (mirrors {@code wishlist.routes.ts} + {@code wishlist.controller.ts}). */
@RestController
@RequestMapping("/api/v1/wishlist")
@Validated
public class WishlistController {

    private final WishlistService wishlistService;

    public WishlistController(WishlistService wishlistService) {
        this.wishlistService = wishlistService;
    }

    @GetMapping
    public ResponseEntity<ApiResponse<List<WishlistItemPublic>>> getWishlist(
            @AuthenticationPrincipal CurrentUser user) {
        CurrentUser current = requireUser(user);
        return ResponseEntity.ok(
                ApiResponse.success(wishlistService.getWishlist(current.userId()), "Wishlist retrieved successfully"));
    }

    @PostMapping("/{productId}")
    public ResponseEntity<ApiResponse<List<WishlistItemPublic>>> addToWishlist(
            @PathVariable String productId, @AuthenticationPrincipal CurrentUser user) {
        CurrentUser current = requireUser(user);
        List<WishlistItemPublic> wishlist = wishlistService.addToWishlist(current.userId(), productId);
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse.success(wishlist, "Item added to wishlist"));
    }

    @DeleteMapping("/{productId}")
    public ResponseEntity<ApiResponse<List<WishlistItemPublic>>> removeFromWishlist(
            @PathVariable String productId, @AuthenticationPrincipal CurrentUser user) {
        CurrentUser current = requireUser(user);
        List<WishlistItemPublic> wishlist = wishlistService.removeFromWishlist(current.userId(), productId);
        return ResponseEntity.ok(ApiResponse.success(wishlist, "Item removed from wishlist"));
    }

    @DeleteMapping
    public ResponseEntity<ApiResponse<List<WishlistItemPublic>>> clearWishlist(
            @AuthenticationPrincipal CurrentUser user) {
        CurrentUser current = requireUser(user);
        List<WishlistItemPublic> wishlist = wishlistService.clearWishlist(current.userId());
        return ResponseEntity.ok(ApiResponse.success(wishlist, "Wishlist cleared successfully"));
    }

    private CurrentUser requireUser(CurrentUser user) {
        if (user == null) {
            throw ApiException.unauthorized("Authentication required", "UNAUTHENTICATED");
        }
        return user;
    }
}
